freestyle-cache/shared/src/main/scala/cache.scala

```diff
@@ -0,0 +1,117 @@
+package freestyle
+
+import cats.{~>}
+
+package cache {
+
+  class KeyValueProvider[Key, Val] {
+
+    /* CacheM trait is a type-class of functors for which key-value store operations
+     * can be provided.
+     *
+     * We assume that the actual store is too big or remote to allow for general
+     * operations over all values, or to search or to iterate over all keys
+     */
+    @free sealed trait CacheM[F[_]] {
+
+      // Gets the value associated to a key, if there is one */
+      def get(key: Key): FreeS.Par[F, Option[Val]]
+
+      // Sets the value of a key to a newValue.
+      def put(key: Key, newVal: Val): FreeS.Par[F, Unit]
```

Comment:
Sounds good, since you seem to have more insight, would you mind creating one and tagging it as such? thanks! @anamariamv Is getting started with Freestyle and she could take on that task as well.